<a href="https://colab.research.google.com/github/bbchen33/Data-Science/blob/master/Genome_data_wrangling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Data wrangling for genome and genes from an unfavored organism - *Clostridium thermocellum*
First, the annotated genome is downloaded from NCBI.

In [32]:
!wget ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/184/925/GCF_000184925.1_ASM18492v1/GCF_000184925.1_ASM18492v1_genomic.gbff.gz

--2019-10-18 03:28:56--  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/184/925/GCF_000184925.1_ASM18492v1/GCF_000184925.1_ASM18492v1_genomic.gbff.gz
           => ‘GCF_000184925.1_ASM18492v1_genomic.gbff.gz’
Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.10, 2607:f220:41e:250::12
Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.10|:21... connected.
Logging in as anonymous ... Logged in!
==> SYST ... done.    ==> PWD ... done.
==> TYPE I ... done.  ==> CWD (1) /genomes/all/GCF/000/184/925/GCF_000184925.1_ASM18492v1 ... done.
==> SIZE GCF_000184925.1_ASM18492v1_genomic.gbff.gz ... 2437250
==> PASV ... done.    ==> RETR GCF_000184925.1_ASM18492v1_genomic.gbff.gz ... done.
Length: 2437250 (2.3M) (unauthoritative)

GCF_000184925.1_ASM 100%[===================>]   2.32M  3.70MB/s    in 0.6s    

2019-10-18 03:28:59 (3.70 MB/s) - ‘GCF_000184925.1_ASM18492v1_genomic.gbff.gz’ saved [2437250]



In [35]:
!ls

GCF_000184925.1_ASM18492v1_genomic.gbff  sample_data


In [0]:
!gunzip GCF_000184925.1_ASM18492v1_genomic.gbff.gz

In [37]:
!head -100 GCF_000184925.1_ASM18492v1_genomic.gbff

LOCUS       NC_017304            3561619 bp    DNA     circular CON 03-SEP-2017
DEFINITION  Hungateiclostridium thermocellum DSM 1313, complete sequence.
ACCESSION   NC_017304
VERSION     NC_017304.1
DBLINK      BioProject: PRJNA224116
            BioSample: SAMN00713572
            Assembly: GCF_000184925.1
KEYWORDS    GSC:MIGS:2.1; RefSeq.
SOURCE      Hungateiclostridium thermocellum DSM 1313
  ORGANISM  Hungateiclostridium thermocellum DSM 1313
            Bacteria; Firmicutes; Clostridia; Clostridiales;
            Hungateiclostridiaceae; Hungateiclostridium.
REFERENCE   1  (bases 1 to 3561619)
  AUTHORS   Feinberg,L., Foden,J., Barrett,T., Davenport,K.W., Bruce,D.,
            Detter,C., Tapia,R., Han,C., Lapidus,A., Lucas,S., Cheng,J.F.,
            Pitluck,S., Woyke,T., Ivanova,N., Mikhailova,N., Land,M.,
            Hauser,L., Argyros,D.A., Goodwin,L., Hogsett,D. and Caiazza,N.
  TITLE     Complete genome sequence of the cellulolytic thermophile
            Clostridium thermoce

In [52]:
table = pd.read_table('GCF_000184925.1_ASM18492v1_genomic.gbff', header = None)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: FutureWarning: read_table is deprecated, use read_csv instead, passing sep='\t'.
  """Entry point for launching an IPython kernel.


In [53]:
table.head()

0
0  LOCUS       NC_017304            3561619 bp   ...
1  DEFINITION  Hungateiclostridium thermocellum D...
2                              ACCESSION   NC_017304
3                            VERSION     NC_017304.1
4                DBLINK      BioProject: PRJNA224116

In [57]:
table[table[0].str.contains('Annotation Date')].index

Int64Index([96], dtype='int64')

In [58]:
table.head(150)

0
0    LOCUS       NC_017304            3561619 bp   ...
1    DEFINITION  Hungateiclostridium thermocellum D...
2                                ACCESSION   NC_017304
3                              VERSION     NC_017304.1
4                  DBLINK      BioProject: PRJNA224116
5                              BioSample: SAMN00713572
6                            Assembly: GCF_000184925.1
7                    KEYWORDS    GSC:MIGS:2.1; RefSeq.
8    SOURCE      Hungateiclostridium thermocellum D...
9      ORGANISM  Hungateiclostridium thermocellum D...
10               Bacteria; Firmicutes; Clostridia; ...
11               Hungateiclostridiaceae; Hungateicl...
12                 REFERENCE   1  (bases 1 to 3561619)
13     AUTHORS   Feinberg,L., Foden,J., Barrett,T.,...
14               Detter,C., Tapia,R., Han,C., Lapid...
15               Pitluck,S., Woyke,T., Ivanova,N., ...
16               Hauser,L., Argyros,D.A., Goodwin,L...
17     TITLE     Complete genome sequence of the ce...
18                    Clostridium thermocellum DSM1313
19     JOURNAL   J. Bacteriol. 193 (11), 2906-2907 ...
20                                   PUBMED   21460082
21                 REFERENCE   2  (bases 1 to 3561619)
22     AUTHORS   Lucas,S., Copeland,A., Lapidus,A.,...
23               Pitluck,S., Davenport,K., Detter,J...
24               Hauser,L., Kyrpides,N., Ivanova,N....
25                  Foden,J., Feinberg,L. and Woyke,T.
26             CONSRTM   US DOE Joint Genome Institute
27     TITLE     Complete sequence of Clostridium t...
28                               JOURNAL   Unpublished
29                 REFERENCE   3  (bases 1 to 3561619)
..                                                 ...
120                     ##Genome-Annotation-Data-END##
121                         COMPLETENESS: full length.
122           FEATURES             Location/Qualifiers
123                         source          1..3561619
124                       /organism="Hungateiclostr...
125                            /mol_type="genomic DNA"
126                                 /strain="DSM 1313"
127                       /culture_collection="DSM:...
128                            /db_xref="taxon:637887"
129                            /collection_date="1926"
130                       /collected_by="Viljoen et...
131                          gene            212..1543
132                       /locus_tag="CLO1313_RS00010"
133                       /old_locus_tag="Clo1313_0...
134                          CDS             212..1543
135                       /locus_tag="CLO1313_RS00010"
136                       /old_locus_tag="Clo1313_0...
137                       /inference="COORDINATES: ...
138                       sequence:RefSeq:WP_020457...
139                       /note="Derived by automat...
140                       gene prediction method: P...
141                                     /codon_start=1
142                                   /transl_table=11
143                       /product="chromosomal rep...
144                       /protein_id="WP_003516465.1"
145                       /translation="MNTQLNEIWQK...
146                       AVPAEFNKGILESRYQTLIKNAIKQ...
147                       PLSVLNPKYTFDTFVIGNSNRFAHA...
148                       GHYILEQNSSQRVLYVSSEKFTNEL...
149                       ERTEEEFFHTFNALYEANKQIILSS...

[150 rows x 1 columns]

It looks like down to row 130 is just the background information. Let's remove them.

In [0]:
table = table.iloc[131:]

In [64]:
table.head(30)

0
131                          gene            212..1543
132                       /locus_tag="CLO1313_RS00010"
133                       /old_locus_tag="Clo1313_0...
134                          CDS             212..1543
135                       /locus_tag="CLO1313_RS00010"
136                       /old_locus_tag="Clo1313_0...
137                       /inference="COORDINATES: ...
138                       sequence:RefSeq:WP_020457...
139                       /note="Derived by automat...
140                       gene prediction method: P...
141                                     /codon_start=1
142                                   /transl_table=11
143                       /product="chromosomal rep...
144                       /protein_id="WP_003516465.1"
145                       /translation="MNTQLNEIWQK...
146                       AVPAEFNKGILESRYQTLIKNAIKQ...
147                       PLSVLNPKYTFDTFVIGNSNRFAHA...
148                       GHYILEQNSSQRVLYVSSEKFTNEL...
149                       ERTEEEFFHTFNALYEANKQIILSS...
150                       IAILRKKAQLENLTVPNEVIVFIAD...
151                       EALKDILSANKAKVLNCTTIQEAVA...
152                       TEMSLPKIGEEFGGRDHTTVIHACE...
153                         gene            1793..2893
154                       /locus_tag="CLO1313_RS00015"
155                       /old_locus_tag="Clo1313_0...
156                         CDS             1793..2893
157                       /locus_tag="CLO1313_RS00015"
158                       /old_locus_tag="Clo1313_0...
159                       /inference="COORDINATES: ...
160                       sequence:RefSeq:WP_003513...

I only want to keep the locus tags and the actual amino acid sequence so I'll delete the rest of the rows.

In [0]:
rows_to_remove_list = ['gene','CDS','/inference','sequence:','/note','gene prediction','/codon_start','transl_table','/product=','protein_id']

In [0]:
for item in rows_to_remove_list:
  table = table[~table[0].str.contains(item)]

In [69]:
table.head(50)

0
132                       /locus_tag="CLO1313_RS00010"
133                       /old_locus_tag="Clo1313_0...
135                       /locus_tag="CLO1313_RS00010"
136                       /old_locus_tag="Clo1313_0...
145                       /translation="MNTQLNEIWQK...
146                       AVPAEFNKGILESRYQTLIKNAIKQ...
147                       PLSVLNPKYTFDTFVIGNSNRFAHA...
148                       GHYILEQNSSQRVLYVSSEKFTNEL...
149                       ERTEEEFFHTFNALYEANKQIILSS...
150                       IAILRKKAQLENLTVPNEVIVFIAD...
151                       EALKDILSANKAKVLNCTTIQEAVA...
152                       TEMSLPKIGEEFGGRDHTTVIHACE...
154                       /locus_tag="CLO1313_RS00015"
155                       /old_locus_tag="Clo1313_0...
157                       /locus_tag="CLO1313_RS00015"
158                       /old_locus_tag="Clo1313_0...
167                       /translation="MKIVCSKEQLM...
168                       GNDLELGIECLIDADILEKGSIVLN...
169                       HFELRGMPSDSFPSLPSIEKENMIK...
170                       AGNEITFVSIDGFRMALRKNFNNEG...
171                       NQILFEIGNCKVVSRLLEGEYLNYK...
172                       KYPVKFNIIDDKIIITSNTEIGAVR...
173                       DIYFNSSVGPCTIRPLEGDSFAYMI...
175                       /locus_tag="CLO1313_RS00020"
176                       /old_locus_tag="Clo1313_0...
178                       /locus_tag="CLO1313_RS00020"
179                       /old_locus_tag="Clo1313_0...
188                       /translation="MENIKINTEFI...
189                          ELRRGRKLRTGDIVEINDKKFQIV"
191                       /locus_tag="CLO1313_RS00025"
192                       /old_locus_tag="Clo1313_0...
194                       /locus_tag="CLO1313_RS00025"
195                       /old_locus_tag="Clo1313_0...
204                       /translation="MYIDRILLKNF...
205                       LCASGRSHRTSKDTELVNIDGTGFS...
206                       KKIGNLMGNLLAVIFSPEDILIIKE...
207                       RNMLLKEIQYKRNLLDTLEVWDYKM...
208                       EIMEIKYSPSVDLHDLSNPSEIKNE...
209                       LNLKMFGSQGQQRTSLLSLKLAEIE...
210                       RNVQTFITCTDKELFENRNFGDNLY...
212                       /locus_tag="CLO1313_RS00030"
213                       /old_locus_tag="Clo1313_0...
215                       /locus_tag="CLO1313_RS00030"
216                       /old_locus_tag="Clo1313_0...
225                       /translation="MFLHIGGDRVV...
226                       QTVSSDIPKSFIITETDKKSIIYLS...
227                                                KG"
230                       /locus_tag="CLO1313_RS00035"
231                       /old_locus_tag="Clo1313_0...
234                       /locus_tag="CLO1313_RS00035"